### PR TITLE
Add miner stats page

### LINKED ISSUE #9: New dashboard for per-miner RSR
**What we want**

We should add a new dashboard to show per-miner RSR as we've described in [spark-stats issue #107](https://github.com/filecoin-station/spark-stats/issues/106)

_A new Observable dashboard showing per-miner stats. At the top, the dashboard should provide a text field for entering the miner ID. Then, there should be a chart showing how this miner's RSR evolved over time. (We will be adding more charts incrementally over time.)_

### PR DESCRIPTION
**Pull request by pyropy**
- Adds miner stats page with per-miner RSR summary visualization 
- Closes #9

### CODE REVIEW

**Comment by pyropy** [comment]
I have tried to create a parametrized data loader and parametrized pages with no success. I don't know is there a cleaner way to dynamically load data to observablehq page?

**Comment by juliangruber** [comment]
I'm not aware of parametrized data loaders being a think in observable (for caching reasons) but I might be wrong

**Comment by juliangruber** [comment]
I tried with the first ID from the miner table but got that error:

Screenshot 2024-12-04 at 10 17 26

**Comment by juliangruber** [comment]
Looks like it's a CORS issue:

```
[Error] Unhandled Promise Rejection: TypeError: undefined is not an object (evaluating 'events.filter')
	LineGraph (line-graph.js:6)
	(anonymous function) (stdlib.js:319)
	(anonymous function) (stdlib.js:313)
[Error] Origin http://127.0.0.1:3000 is not allowed by Access-Control-Allow-Origin. Status code: 200
[Error] Fetch API cannot load https://stats.filspark.com/miner/f03156617/retrieval-success-rate/summary?from=2024-06-07&to=2024-12-03 due to access control checks.
[Error] Failed to load resource: Origin http://127.0.0.1:3000 is not allowed by Access-Control-Allow-Origin. Status code: 200 (summary, line 0)
```

**Review by juliangruber — CHANGES_REQUESTED** [review]
Let's take inspiration from this dashboard: https://filecoin-project.github.io/filecoin-storage-providers-market/

1. In the miner table, each entry links to a miner detail page
2. The miner id is part of the URL linked to
3. It looks like there is a dynamic data source at play: https://github.com/filecoin-project/filecoin-storage-providers-market/blob/46d85042282e2a722978c9131659cf5a1a7e96ef/src/data/%5Bprovider%5D_daily_metrics.csv.sh
4. There's a dynamic markdown document here: https://github.com/filecoin-project/filecoin-storage-providers-market/blob/main/src/provider/%5Bprovider%5D.md
5. There's a build step that fetches all known miner IDs: https://github.com/filecoin-project/filecoin-storage-providers-market/blob/main/observablehq.config.js#L7

**Comment by juliangruber** [comment]
@pyropy I've converted this PR to a draft, to signal that it isn't ready to be merged yet

**Comment by pyropy** [comment]
> Looks like it's a CORS issue:
> 
> ```
> [Error] Unhandled Promise Rejection: TypeError: undefined is not an object (evaluating 'events.filter')
> 	LineGraph (line-graph.js:6)
> 	(anonymous function) (stdlib.js:319)
> 	(anonymous function) (stdlib.js:313)
> [Error] Origin http://127.0.0.1:3000 is not allowed by Access-Control-Allow-Origin. Status code: 200
> [Error] Fetch API cannot load https://stats.filspark.com/miner/f03156617/retrieval-success-rate/summary?from=2024-06-07&to=2024-12-03 due to access control checks.
> [Error] Failed to load resource: Origin http://127.0.0.1:3000 is not allowed by Access-Control-Allow-Origin. Status code: 200 (summary, line 0)
> ```


> Looks like it's a CORS issue:
> 
> ```
> [Error] Unhandled Promise Rejection: TypeError: undefined is not an object (evaluating 'events.filter')
> 	LineGraph (line-graph.js:6)
> 	(anonymous function) (stdlib.js:319)
> 	(anonymous function) (stdlib.js:313)
> [Error] Origin http://127.0.0.1:3000 is not allowed by Access-Control-Allow-Origin. Status code: 200
> [Error] Fetch API cannot load https://stats.filspark.com/miner/f03156617/retrieval-success-rate/summary?from=2024-06-07&to=2024-12-03 due to access control checks.
> [Error] Failed to load resource: Origin http://127.0.0.1:3000 is not allowed by Access-Control-Allow-Origin. Status code: 200 (summary, line 0)
> ```

Yeah that's weird, I've assumed that new API endpoint was not deployed so I just was fetching from local API instead. I'm going to check that out.

**Comment by juliangruber** [comment]
No need to check out the CORS issue since we'll move the fetching to a server-side data source again (like the other data sources we have)

**Comment by pyropy** [comment]
> Let's take inspiration from this dashboard: https://filecoin-project.github.io/filecoin-storage-providers-market/
> 
>     1. In the miner table, each entry links to a miner detail page
> 
>     2. The miner id is part of the URL linked to
> 
>     3. It looks like there is a dynamic data source at play: https://github.com/filecoin-project/filecoin-storage-providers-market/blob/46d85042282e2a722978c9131659cf5a1a7e96ef/src/data/%5Bprovider%5D_daily_metrics.csv.sh
> 
>     4. There's a dynamic markdown document here: https://github.com/filecoin-project/filecoin-storage-providers-market/blob/main/src/provider/%5Bprovider%5D.md
> 
>     5. There's a build step that fetches all known miner IDs: https://github.com/filecoin-project/filecoin-storage-providers-market/blob/main/observablehq.config.js#L7

Yeah this is what I've tried to do at first but honestly I was lacking a good example (docs do describe it but I haven't managed to do it following  Also, I decided to do it on the client side as we have this text input that is used to query for miner id.

What comes to my mind is to maybe redirect users to miner page once they submit miner id they want to query. Second option would be to list miners on some page and have them click on miner id (similar to example you've posted in the comment).

**Comment by juliangruber** [comment]
> Yeah this is what I've tried to do at first but honestly I was lacking a good example (docs do describe it but I haven't managed to do it following

I'm happy we have that working example available 👍  Otherwise I also didn't find good docs for that.

> Also, I decided to do it on the client side as we have this text input that is used to query for miner id.

I don't think we should have a text input there, in the RSR table at the bottom of https://dashboard.filspark.com there's a search functionality, and from there we should link to the miner detail page (without text input).

Does that work for you?

**Comment by pyropy** [comment]
> > Yeah this is what I've tried to do at first but honestly I was lacking a good example (docs do describe it but I haven't managed to do it following
> 
> I'm happy we have that working example available 👍 Otherwise I also didn't find good docs for that.
> 
> > Also, I decided to do it on the client side as we have this text input that is used to query for miner id.
> 
> I don't think we should have a text input there, in the RSR table at the bottom of https://dashboard.filspark.com there's a search functionality, and from there we should link to the miner detail page (without text input).
> 
> Does that work for you?

Yeah that sounds good to me!

File: src/data/[minerId]-spark-rsr-summary.json.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+const start = '2024-04-07';
+const end = getDateXDaysAgo(1);
+
+const summaries = await jsonFetcher(
+  `https://stats.filspark.com/miner/${minerId}/retrieval-success-rate/summary?from=${start}&to=${end}`
+);
+
+
+process.stdout.write(JSON.stringify(summaries));
```
Comment (juliangruber) [inline src/data/[minerId]-spark-rsr-summary.json.js:19]:
```suggestion
const summary = await jsonFetcher(
  `https://stats.filspark.com/miner/${minerId}/retrieval-success-rate/summary?from=${start}&to=${end}`
);


process.stdout.write(JSON.stringify(summary));
```

File: src/miner/[minerId].md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+<div class="hero">
+  <body><a href="/"><img src="../media/spark-logomark-blue-with-bbox.png" alt="Spark Logo" width="300" /></a><body>
+    <h2>Dashboard Beta</h2>
+    <body><a href="https://filspark.com/dashboard" target="_blank" rel="noopener noreferrer">(Click here for Legacy Spark Grafana Dashboard)</a><body>
+</div>
+
+<h4>Miner Spark RSR Summary</h4>
+<body>This section shows the miner Spark Retrieval Success Rate Score summary. You can adjust the date range. Records start on the 7th April 2024.</body>
```
Comment (juliangruber) [inline src/miner/[minerId].md:14]: Nothing happens when I adjust the date range 🤔

File: src/miner/[minerId].md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+
+```js
+const start = view(Inputs.date({label: "Start", value: getDateXDaysAgo(180) }));
```
Comment (juliangruber) [inline src/miner/[minerId].md:18]:
```suggestion
const start = view(Inputs.date({label: "Start", value: "2024-04-07" }));
```

I think it makes sense to show all data by default

Comment (juliangruber) [reply src/miner/[minerId].md:18]: But I'm not sure

File: src/miner/[minerId].md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+const end = view(Inputs.date({label: "End", value: getDateXDaysAgo(1) }));
+```
+
+<h3>Data for Miner ID: ${observable.params.minerId}</h3>
```
Comment (juliangruber) [inline src/miner/[minerId].md:22]:
```suggestion
<h3>Stats for ${observable.params.minerId}</h3>
```

Data is very generic, plus we're not really supposed to call providers miners any more.

**Review by juliangruber — CHANGES_REQUESTED** [review]
Almost there 👏

File: src/miner/[minerId].md (hunk 3 above)
Comment (pyropy) [reply src/miner/[minerId].md:18]: I've seen that main dashboard shows data for the past 180 days so I went with it.

File: src/miner/[minerId].md (hunk 4 above)
Comment (pyropy) [reply src/miner/[minerId].md:22]: Thanks for heads up, I might update the route too (change from /miner to /provider)?

File: src/miner/[minerId].md (hunk 2 above)
Comment (pyropy) [reply src/miner/[minerId].md:14]: I think there's some strange behavior if there's no data for provider in certain range. 

If we compare these two providers [f010479](https://add-per-miner-rsr-visualizat.spark-dashboard.pages.dev/miner/f010479) and [f03156617](https://add-per-miner-rsr-visualizat.spark-dashboard.pages.dev/miner/f03156617) earliest entry in database for f010479 is on 7th April 2024 but f03156617 it's 29th July 2024.

**Comment by pyropy** [comment]
Should we potentially enable tips for this chart? Maybe it's too noisy. 

Screenshot 2024-12-04 at 14 35 23

**Comment by bajtos** [comment]
I know almost nothing about the Observable framework, so I will leave the review & final approval in your hands, @juliangruber.

Re: server-side or client-side data fetching.

Do we have a rough understanding of how much the build will be slowed down by having to fetch data for all miners? We are starting with one request per miner now, but I'd like to incrementally add more charts, maybe up to 5-10 charts per miner. If there are 200 active miners, we must make 1000-5000 requests to build the static data.

I will also want to introduce some live data to the per-miner page to help them diagnose issues. For example:
- What's their miner's Peer ID. 
- What's their IPNI provider ID. 
- What's the IPNI ingestion status. 
- How many deals eligible for retrieval testing we have in our DB. This will change every FIL epoch (30s) in Spark v2.
I guess the diagnostic can go to a different page that will use client-side data fetching, so this part is not relevant to this pull request.

**Comment by bajtos** [comment]
Other than that, the new dashboard looks great!

I propose changing the description of the miner table to invite people to click on the miner id to see per-miner details.

```diff
- The following table shows the Spark RSR values calculated in aggregate for each Filecoin Storage Provider over the past 30 days.
+ The following table shows the Spark RSR values calculated in aggregate for each Filecoin Storage Provider over the past 30 days. Click on a miner_id to view data scoped to this miner.
```

File: src/miner/[minerId].md (hunk 4 above)
Comment (juliangruber) [reply src/miner/[minerId].md]: Yes please

File: src/index.md [hunk 5]
```diff
@@ -238,14 +238,14 @@ ${JSON.stringify(
 <div class="divider"></div>
 
 <h4>Spark Miner RSR Table</h4>
-<body>The following table shows the Spark RSR values calculated in aggregate for each Filecoin Storage Provider over the past 30 days.</body>
+<body>The following table shows the Spark RSR values calculated in aggregate for each Filecoin Storage Provider over the past 30 days. Click on a miner_id to view data scoped to this storage provider.</body>
```
Comment (juliangruber) [inline src/index.md:241]:
```suggestion
<body>The following table shows the Spark RSR values calculated in aggregate for each Filecoin Storage Provider over the past 30 days. Click on a miner id to view stats about this storage provider.</body>
```

(less technical)

**Review by juliangruber — APPROVED** [review]

**Comment by juliangruber** [comment]
> Do we have a rough understanding of how much the build will be slowed down by having to fetch data for all miners? We are starting with one request per miner now, but I'd like to incrementally add more charts, maybe up to 5-10 charts per miner. If there are 200 active miners, we must make 1000-5000 requests to build the static data.

I don't see a problem with that, it's still simpler to handle than dynamic pages. And we can move to client side fetching should it eventually become a problem.


> I will also want to introduce some live data to the per-miner page to help them diagnose issues. For example:
> 
> * What's their miner's Peer ID.
> * What's their IPNI provider ID.

This can be static

> * What's the IPNI ingestion status.
> * How many deals eligible for retrieval testing we have in our DB. This will change every FIL epoch (30s) in Spark v2.
>   I guess the diagnostic can go to a different page that will use client-side data fetching, so this part is not relevant to this pull request.

Sounds good!

**Comment by juliangruber** [comment]
Screenshot 2024-12-04 at 23 06 35

Here are build times for this PR. I tried to find build times for current main, but couldn't, when following the checks. Maybe that is because the last deploy was too long ago?

Anyway, I think 10 minutes is fine for now.

**Comment by juliangruber** [comment]
See https://dash.cloudflare.com/37573110e38849a343d93b727953188f/pages/view/spark-dashboard/faf33bf8-f390-474a-be9c-9a505c6a71c0

**Comment by pyropy** [comment]
Seems like a build failed now. Could someone provide me with access to logs?

**Comment by bajtos** [comment]
> > I will also want to introduce some live data to the per-miner page to help them diagnose issues. For example:
> > 
> > * What's their miner's Peer ID.
> > * What's their IPNI provider ID.
> 
> This can be static

Not really. These Peer IDs can be changed - I learned this after an SP reached out after they changed their Peer ID.

**Comment by juliangruber** [comment]
[spark-dashboard.9666a29e-3a81-4359-9a86-cf8e992a93ee.log](https://github.com/user-attachments/files/18021407/spark-dashboard.9666a29e-3a81-4359-9a86-cf8e992a93ee.log)

@pyropy While we're working on giving you Cloudflare access, here's the logs of the failed build

**Comment by juliangruber** [comment]
Two things:
1. In this case, can you try re-running the GitHub check?
2. It looks like a request failed, which made the whole build fail. Let's add retry logic here (we mostly use https://www.npmjs.com/package/p-retry)

**Comment by pyropy** [comment]
> Two things:
> 
>     1. In this case, can you try re-running the GitHub check?
> 
>     2. It looks like a request failed, which made the whole build fail. Let's add retry logic here (we mostly use https://www.npmjs.com/package/p-retry)

Since the checks are passing and `p-retry` has been added, I guess this can be merged now?

**Pull request merged**